 <a name="1"></a>
# <span style="color:green">Задачи исследования:<span>
    
- проанализировать тарифные планы компании
- проанализировать поведение клиентов компании использующие тарифные планы представленные ниже
- сделать вывод — какой тариф приносит больше денег.

### Описание тарифов

##### Тариф «Смарт»

1. Ежемесячная плата: 550 рублей
2. Включено 500 минут разговора, 50 сообщений и 15 Гб интернет-трафика
3. Стоимость услуг сверх тарифного пакета:
    - минута разговора: 3 рубля («Мегалайн» всегда округляет вверх значения минут и мегабайтов.
    Если пользователь проговорил всего 1 секунду, в тарифе засчитывается целая минута.)
    - сообщение: 3 рубля
    - 1 Гб интернет-трафика: 200 рублей

##### Тариф «Ультра»

1. Ежемесячная плата: 1950 рублей
2. Включено 3000 минут разговора, 1000 сообщений и 30 Гб интернет-трафика
3. Стоимость услуг сверх тарифного пакета:
    - минута разговора: 1 рубль
    - сообщение: 1 рубль
    - 1 Гб интернет-трафика: 150 рублей

# План 

## 1. [Осмотр и изучение данных](#1)

## 2. [Предобработка данных](#2)

- исправление ошибок в данных
- проверка и предварительный анализ

## 3. [Подготовка, очистка и добавление данных](#3)

- индексы и названия столбцов
- работа с дубликатами
- работа с пропущенными значениями
- изменение типов данных
- добавление дополнительных метрик **для каждого пользователя:**
      
      
    - количество сделанных звонков и израсходованных минут разговора по месяцам;
    - количество отправленных сообщений по месяцам;
    - объем израсходованного интернет-трафика по месяцам;
    - помесячную выручку с каждого пользователя (вычтите бесплатный лимит из суммарного количества звонков, сообщений и интернет-трафика; остаток умножьте на значение из тарифного плана).

## 4. [Исследовательский анализ данных](#4)

- Опишите поведение клиентов оператора, исходя из выборки.
- Сколько минут разговора, сколько сообщений и какой объём интернет-трафика требуется пользователям каждого тарифа в месяц?
- Посчитайте среднее количество, дисперсию и стандартное отклонение. Постройте гистограммы. Опишите распределения.(#41)
  
## 5. [Проверка гипотез](#5)

- средняя выручка пользователей тарифов «Ультра» и «Смарт» различается;
- средняя выручка пользователей из Москвы отличается от выручки пользователей из других регионов;
  
## 6. [Выводы и результаты исследования](#6)

# 1. Импорт библиотек и чтение данных

In [239]:
import random
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# Disable warnings in Anaconda
import warnings
warnings.filterwarnings('ignore')

from pylab import rcParams
rcParams['figure.figsize'] = 10,7
pd.set_option('display.max_columns', None)

In [240]:
calls = pd.read_csv('datasets/calls.csv')
internet = pd.read_csv('datasets/internet.csv')
messages = pd.read_csv('datasets/messages.csv')
tariffs = pd.read_csv('datasets/tariffs.csv')
users = pd.read_csv('datasets/users.csv')

### 1.1 Таблица звонков

In [241]:
calls.head(3)

id   call_date  duration  user_id
0  1000_0  2018-07-25      0.00     1000
1  1000_1  2018-08-17      0.00     1000
2  1000_2  2018-06-11      2.85     1000

In [242]:
calls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202607 entries, 0 to 202606
Data columns (total 4 columns):
id           202607 non-null object
call_date    202607 non-null object
duration     202607 non-null float64
user_id      202607 non-null int64
dtypes: float64(1), int64(1), object(2)
memory usage: 6.2+ MB


### 1.2 Таблица интернет трафика

In [243]:
internet.head(3)

id  mb_used session_date  user_id
0  1000_0   112.95   2018-11-25     1000
1  1000_1  1052.81   2018-09-07     1000
2  1000_2  1197.26   2018-06-25     1000

In [244]:
internet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149396 entries, 0 to 149395
Data columns (total 4 columns):
id              149396 non-null object
mb_used         149396 non-null float64
session_date    149396 non-null object
user_id         149396 non-null int64
dtypes: float64(1), int64(1), object(2)
memory usage: 4.6+ MB


### 1.3 Таблица сообщений

In [245]:
messages.head(3)

id message_date  user_id
0  1000_0   2018-06-27     1000
1  1000_1   2018-10-08     1000
2  1000_2   2018-08-04     1000

In [246]:
messages.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123036 entries, 0 to 123035
Data columns (total 3 columns):
id              123036 non-null object
message_date    123036 non-null object
user_id         123036 non-null int64
dtypes: int64(1), object(2)
memory usage: 2.8+ MB


### 1.3 Таблица пользователей

In [247]:
users.head(3)

user_id  age churn_date         city first_name  last_name    reg_date  \
0     1000   52        NaN    Краснодар     Рафаил  Верещагин  2018-05-25   
1     1001   41        NaN       Москва       Иван       Ежов  2018-11-01   
2     1002   59        NaN  Стерлитамак    Евгений  Абрамович  2018-06-17   

  tariff  
0  ultra  
1  smart  
2  smart

In [248]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 8 columns):
user_id       500 non-null int64
age           500 non-null int64
churn_date    38 non-null object
city          500 non-null object
first_name    500 non-null object
last_name     500 non-null object
reg_date      500 non-null object
tariff        500 non-null object
dtypes: int64(2), object(6)
memory usage: 31.4+ KB


### 1.4 Таблица тарифов

In [249]:
tariffs.head()

messages_included  mg_per_month_included  minutes_included  \
0                 50                  15360               500   
1               1000                  30720              3000   

   rub_monthly_fee  rub_per_gb  rub_per_message  rub_per_minute tariff_name  
0              550         200                3               3       smart  
1             1950         150                1               1       ultra

### 1.5 Общая таблица данных

In [250]:
len(users), len(calls), len(messages), len(internet)

(500, 202607, 123036, 149396)

In [251]:
all_calls_user = calls.pivot_table(index='user_id', values='duration')
all_calls_user.head()

duration
user_id          
1000     6.571304
1001     6.552951
1002     6.595253
1003     6.924077
1004     5.234148

In [252]:
all_internet_user = internet.pivot_table(index='user_id', values='mb_used')
all_internet_user.head()

mb_used
user_id            
1000     508.968294
1001     338.187500
1002     314.462114
1003     550.775385
1004     496.798267

In [253]:
all_messages_user = messages.pivot_table(index='user_id', aggfunc='count')
all_messages_user.drop(['id'], axis=1, inplace=True)
all_messages_user.rename({'message_date': 'count_messages'}, axis=1, inplace=True)
all_messages_user.head()

count_messages
user_id                
1000                496
1002                 70
1003                380
1004               1192
1005                552

In [254]:
df = ( 
        users.merge(all_calls_user, how='outer', on='user_id')
             .merge(all_internet_user, how='outer', on='user_id')
             .merge(all_messages_user, how='outer', on='user_id')
      )

In [255]:
df.head()

user_id  age churn_date         city first_name  last_name    reg_date  \
0     1000   52        NaN    Краснодар     Рафаил  Верещагин  2018-05-25   
1     1001   41        NaN       Москва       Иван       Ежов  2018-11-01   
2     1002   59        NaN  Стерлитамак    Евгений  Абрамович  2018-06-17   
3     1003   23        NaN       Москва      Белла   Белякова  2018-08-17   
4     1004   68        NaN  Новокузнецк    Татьяна   Авдеенко  2018-05-14   

  tariff  duration     mb_used  count_messages  
0  ultra  6.571304  508.968294           496.0  
1  smart  6.552951  338.187500             NaN  
2  smart  6.595253  314.462114            70.0  
3  ultra  6.924077  550.775385           380.0  
4  ultra  5.234148  496.798267          1192.0

In [256]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 11 columns):
user_id           500 non-null int64
age               500 non-null int64
churn_date        38 non-null object
city              500 non-null object
first_name        500 non-null object
last_name         500 non-null object
reg_date          500 non-null object
tariff            500 non-null object
duration          492 non-null float64
mb_used           497 non-null float64
count_messages    426 non-null float64
dtypes: float64(3), int64(2), object(6)
memory usage: 46.9+ KB


 <a name="2"></a>
# 2. Предобработка данных и предварительный анализ
[К оглавлению](#1)

### Распределение пользователей по тарифам

In [257]:
df['tariff'].value_counts()

smart    351
ultra    149
Name: tariff, dtype: int64

In [258]:
## дейсвтующих
df[df['churn_date'].isna()]['tariff'].value_counts()

smart    328
ultra    134
Name: tariff, dtype: int64

In [259]:
## прекративших пользоваться услугами
df[df['churn_date'].isna()]['tariff'].value_counts()

smart    328
ultra    134
Name: tariff, dtype: int64

### Звоники длительностью менее 1 минуты

In [260]:
calls[calls['duration'] < 1]

id   call_date  duration  user_id
0         1000_0  2018-07-25      0.00     1000
1         1000_1  2018-08-17      0.00     1000
5         1000_5  2018-11-02      0.00     1000
6         1000_6  2018-10-18      0.00     1000
9         1000_9  2018-08-15      0.00     1000
...          ...         ...       ...      ...
202592  1499_205  2018-10-13      0.00     1499
202593  1499_206  2018-12-26      0.00     1499
202594  1499_207  2018-10-28      0.00     1499
202596  1499_209  2018-12-16      0.00     1499
202602  1499_215  2018-12-26      0.76     1499

[47927 rows x 4 columns]

### Вывод:
**Всего в нашей выборке 500 абонентов:**
- тариф smart 351 пользователя
- тариф ultra 149 пользователя

##### из которых действующих 462:

- тариф smart 328
- тариф ultra 134

##### прекративших пользоваться услугами 38**:
- тариф smart 23
- тариф ultra 15

**Предварительный анализ показал что:**

- Требуется изменить в данных тип данных всех дат в datetime
- Столбец **churn_date** можно изменить на True и False - по желанию
- Опереатор всегда округляет вверх значения минут и мегабайтов. Если пользователь проговорил всего 1 секунду, в тарифе засчитывается целая минута. Значит требуется округлить значения меньше минуты 
- Интернет трафик по тарифу расчитывается в Гб а в таблице Мб. Обратить внимание при обобщении трафика для каждого usera 

<a name="3"></a>
# 3. Подготовка и очистка данных
[К оглавлению](#1)

### Изменяем тип данных в таблицах

In [261]:
def get_datetime(data, columns):
    for col in columns:
        data[col] = data[col].astype('datetime64[ns]')
    return data

In [262]:
get_datetime(calls, ['call_date'])
get_datetime(internet, ['session_date'])
get_datetime(messages, ['message_date'])
get_datetime(users, ['reg_date', 'churn_date'])

user_id  age churn_date         city first_name  last_name   reg_date  \
0       1000   52        NaT    Краснодар     Рафаил  Верещагин 2018-05-25   
1       1001   41        NaT       Москва       Иван       Ежов 2018-11-01   
2       1002   59        NaT  Стерлитамак    Евгений  Абрамович 2018-06-17   
3       1003   23        NaT       Москва      Белла   Белякова 2018-08-17   
4       1004   68        NaT  Новокузнецк    Татьяна   Авдеенко 2018-05-14   
..       ...  ...        ...          ...        ...        ...        ...   
495     1495   65        NaT      Иркутск  Авксентий      Фокин 2018-08-28   
496     1496   36        NaT      Вологда     Трифон     Блохин 2018-01-27   
497     1497   32        NaT    Челябинск   Каролина     Агеева 2018-10-09   
498     1498   68 2018-10-25  Владикавказ   Всеволод    Акимчин 2018-07-19   
499     1499   35        NaT        Пермь     Гектор   Корнилов 2018-09-27   

    tariff  
0    ultra  
1    smart  
2    smart  
3    ultra  
4    ultra  
..     ...  
495  ultra  
496  smart  
497  smart  
498  smart  
499  smart  

[500 rows x 8 columns]

### Округлим данные по звонкам и заменим нули на 1

In [263]:
# Смотрим сколько данных меньше 1 минуты
calls[calls['duration'] < 1]

id  call_date  duration  user_id
0         1000_0 2018-07-25      0.00     1000
1         1000_1 2018-08-17      0.00     1000
5         1000_5 2018-11-02      0.00     1000
6         1000_6 2018-10-18      0.00     1000
9         1000_9 2018-08-15      0.00     1000
...          ...        ...       ...      ...
202592  1499_205 2018-10-13      0.00     1499
202593  1499_206 2018-12-26      0.00     1499
202594  1499_207 2018-10-28      0.00     1499
202596  1499_209 2018-12-16      0.00     1499
202602  1499_215 2018-12-26      0.76     1499

[47927 rows x 4 columns]

In [264]:
# Приводим значения в порядок
calls['duration'] = calls['duration'].apply(np.ceil).astype('int').replace(0, 1)
calls.head()

id  call_date  duration  user_id
0  1000_0 2018-07-25         1     1000
1  1000_1 2018-08-17         1     1000
2  1000_2 2018-06-11         3     1000
3  1000_3 2018-09-21        14     1000
4  1000_4 2018-12-15         6     1000

### Округлим данные интерент сессий, заменим нули на 1

In [265]:
# Смотрим данные по интернет сессиям
internet[internet['mb_used'] < 1]

id  mb_used session_date  user_id
11       1000_11      0.0   2018-08-28     1000
38       1000_38      0.0   2018-11-27     1000
46       1000_46      0.0   2018-06-30     1000
67       1000_67      0.0   2018-06-16     1000
72       1000_72      0.0   2018-06-15     1000
...          ...      ...          ...      ...
149380  1499_141      0.0   2018-12-11     1499
149381  1499_142      0.0   2018-11-01     1499
149383  1499_144      0.0   2018-11-27     1499
149386  1499_147      0.0   2018-10-21     1499
149393  1499_154      0.0   2018-10-27     1499

[19699 rows x 4 columns]

In [266]:
# Приводим значения в порядок
internet['mb_used'] = internet['mb_used'].apply(np.ceil).astype('int').replace(0, 1)
internet

id  mb_used session_date  user_id
0         1000_0      113   2018-11-25     1000
1         1000_1     1053   2018-09-07     1000
2         1000_2     1198   2018-06-25     1000
3         1000_3      551   2018-08-22     1000
4         1000_4      303   2018-09-24     1000
...          ...      ...          ...      ...
149391  1499_152      319   2018-10-03     1499
149392  1499_153      491   2018-12-14     1499
149393  1499_154        1   2018-10-27     1499
149394  1499_155     1247   2018-11-26     1499
149395  1499_156      545   2018-10-26     1499

[149396 rows x 4 columns]

### Добавление дополнительных метрик для пользователей:
- количество сделанных звонков и израсходованных минут разговора по месяцам;
- количество отправленных сообщений по месяцам;
- объем израсходованного интернет-трафика по месяцам;
- помесячную выручку с каждого пользователя (вычтите бесплатный лимит из суммарного количества звонков, сообщений и интернет-трафика; остаток умножьте на значение из тарифного плана).